### PR TITLE
RavenDB-17608 Query execution planner - UnaryMatches

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17608

### Additional description

Before optimization query like this:
`where x < 5 and y > 10 and z <=32`

were evaluated by the intersection of each set which is very costly.
Now we are merging it into one query stack.
For example
Check which entries have x < 5 in the entire set of terms `x`. 
Knowing which ones satisfy this condition, we can check which ones satisfy the `y>10` condition inside the previous set of matches, and so on.

Also we are transforming ` x >0 and  x < 10` into `between`

We check 

### Type of change

- Optimization

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Queries/BoostingMatch.cs [hunk 1]
```diff
@@ -157,7 +158,10 @@ private void UnlikelyGrowBuffer(int size)
             var bufferHandler = _searcher.Allocator.Allocate(size, out var buffer);
             
             // Ensure we copy the content and then switch the buffers. 
-            new Span<long>(_buffer, _bufferIdx).CopyTo(new Span<long>(buffer.Ptr, size));
+            new Span<long>(_buffer, _bufferIdx).CopyTo(new Span<long>(buffer.Ptr, size));// == false)
+           // {
+           //     throw new Exception("Failed to copy the content of the buffer");
+           // }
```
Comment (arekpalinski) [inline src/Corax/Queries/BoostingMatch.cs]: Remove commented code please

File: src/Corax/Queries/MemoizationMatch.cs [hunk 2]
```diff
@@ -22,6 +22,12 @@ internal MemoizationMatch(in MemoizationMatchProvider<TInner> inner)
              
         public int Fill(Span<long> matches)
         {
+            if (_inner.ReplayCounter == 1)
+            {
+                //We tried to pass around memoized version of TInner but there is no need to persist it
```
Comment (arekpalinski) [inline src/Corax/Queries/MemoizationMatch.cs:27]: memoized -> memorized

File: src/Corax/Queries/TermProviders/TermProvider.ExistsWithComparer.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+            term = TermMatch.CreateEmpty();
+            return false;
+        }
+
+        public bool GetNextTerm(out ReadOnlySpan<byte> term)
+        {
+            while (_iterator.MoveNext(out Span<byte> termSlice, out var _))
+            {
+                // This shouldnt happen.
```
Comment (arekpalinski) [inline src/Corax/Queries/TermProviders/TermProvider.ExistsWithComparer.cs]: Let's add Debug assert for that

File: src/Corax/Queries/UnaryMatch.Between.cs [hunk 4]
```diff
@@ -194,19 +193,18 @@ unsafe partial struct UnaryMatch<TInner, TValueType>
...
                     var reader = searcher.GetReaderFor(currentMatches[i]);
                     var read = reader.Read(match._fieldId, out var resultX);
-                    if ((read && leftSideComparer!.Compare(currentType1, resultX) && rightSideComparer!.Compare(currentType2, resultX)) == false)
+                    if (read && leftSideComparer!.Compare(currentType1, resultX) && rightSideComparer!.Compare(currentType2, resultX))
                     {
-                        // We found a match.
-                        currentMatches[currentIdx] = currentMatches[storeIdx];
-                        storeIdx++;
+                        // We found a match so we have to skip it.
+                       continue;
```
Comment (arekpalinski) [inline src/Corax/Queries/UnaryMatch.Between.cs:204]: Just to make sure  - I see you inverted the condition but now you no longer do `storeIdx++;` isn't it necessary?

File: src/Corax/Queries/UnaryMatch.Ordinal.cs [hunk 5]
```diff
@@ -36,6 +37,36 @@ private static int AndWith(ref UnaryMatch<TInner, TValueType> match, Span<long>
             return result;
         }
 
+        // private static int FillFuncSequenceAnyBasedOnTerms<TComparer>(ref UnaryMatch<TInner, TValueType> match, Span<long> matches)
```
Comment (arekpalinski) [inline src/Corax/Queries/UnaryMatch.Ordinal.cs]: Please remove commented code

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 6]
```diff
@@ -113,16 +113,28 @@ protected static int LuceneGetPageSize(IndexSearcher searcher, long pageSize)
...
 
-    protected static int CoraxGetPageSize(global::Corax.IndexSearcher searcher, int bufferSize, IndexQueryServerSide query)
+    protected static int CoraxGetPageSize(global::Corax.IndexSearcher searcher, int pageSize, IndexQueryServerSide query, bool isBoolean = false)
     {
-        var size = searcher.NumberOfEntries;
-        if (size > int.MaxValue)
+        var numberOfEntries = searcher.NumberOfEntries;
+        
+        if (numberOfEntries == 0)
+            return 2 << 4;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs]: Readability issue -  why this way?

File: src/Corax/Queries/MemoizationMatch.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Corax/Queries/MemoizationMatch.cs:27]: https://en.wikipedia.org/wiki/Memoization

File: src/Corax/Queries/BoostingMatch.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/Queries/BoostingMatch.cs]: done

File: src/Corax/Queries/TermProviders/TermProvider.ExistsWithComparer.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Corax/Queries/TermProviders/TermProvider.ExistsWithComparer.cs]: done

File: src/Corax/Queries/UnaryMatch.Between.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Corax/Queries/UnaryMatch.Between.cs:204]: `storeIdx` is the index on the set we want to return, using `i` to iterate over the set where the candidates are. 

Please notice this is NotBetween.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs]: Done

**Comment by arekpalinski** [comment]
@maciejaszyk Please check Debug.Assert failure:

https://github.com/ravendb/ravendb/runs/7018216852?check_suite_focus=true#step:5:94

**Comment by maciejaszyk** [comment]
> @maciejaszyk Please check Debug.Assert failure:
> 
> https://github.com/ravendb/ravendb/runs/7018216852?check_suite_focus=true#step:5:94

fixed

**Review by arekpalinski — APPROVED** [review]

File: src/Corax/Queries/MemoizationMatch.cs (hunk 2 above)
Comment (redknightlois) [reply src/Corax/Queries/MemoizationMatch.cs:27]: The correct word is Memoization

File: src/Corax/IndexWriter.cs [hunk 7]
```diff
@@ -305,6 +306,10 @@ void Insert(ReadOnlySpan<byte> value)
 
             void AnalyzeInsert(Span<byte> wordsBuffer, Span<Token> tokens, ReadOnlySpan<byte> value)
             {
+                if (wordsBuffer.Length < value.Length)
+                {
+                    throw new Exception($"Passed value with length: {value.Length}. Contains {Encodings.Utf8.GetString(value)} but buffer is only {wordsBuffer.Length}");
```
Comment (ayende) [inline src/Corax/IndexWriter.cs:311]: Let's not throw raw `Exception`

File: src/Corax/Queries/MemoizationMatch.Provider.cs [hunk 8]
```diff
@@ -6,9 +6,12 @@
 namespace Corax.Queries
 {
 
-    public unsafe struct MemoizationMatchProvider<TInner> : IMemoizationMatchSource
+    public unsafe class MemoizationMatchProvider<TInner> : IMemoizationMatchSource
```
Comment (ayende) [inline src/Corax/Queries/MemoizationMatch.Provider.cs:9]: Why is this a class now? Is this instance shared? In what context? What about multi threading?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        Unsafe.SkipInit(out BetweenRight);
+
+        if (operation is UnaryMatchOperation.Equals or UnaryMatchOperation.NotEquals)
+            TermAsString = QueryBuilderHelper.CoraxGetValueAsString(term);
+        else
+            Unsafe.SkipInit(out TermAsString);
+
+        Count = operation is UnaryMatchOperation.Equals or UnaryMatchOperation.NotEquals
+            ? searcher.TermAmount(name, TermAsString)
+            : searcher.GetEntriesAmountInField(name);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs:52]: Shouldn't we set the `if` once?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    {
+        return (lhs._scoreFunction, rhs._scoreFunction) switch
+        {
+            (NullScoreFunction, NullScoreFunction) => true,
+            (ConstantScoreFunction lcsf, ConstantScoreFunction rcsf) => lcsf.Value.AlmostEquals(rcsf.Value),
+            _ => false
+        };
+    }
+
+    public IQueryMatch Materialize()
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxBooleanQuery.cs:85]: Can you discuss this with me, I want to understand your logic here.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 11]
```diff
@@ -113,16 +113,26 @@ protected static int LuceneGetPageSize(IndexSearcher searcher, long pageSize)
...
 
-    protected static int CoraxGetPageSize(global::Corax.IndexSearcher searcher, int bufferSize, IndexQueryServerSide query)
+    protected static int CoraxGetPageSize(global::Corax.IndexSearcher searcher, int pageSize, IndexQueryServerSide query, bool isBoolean = false)
     {
-        var size = searcher.NumberOfEntries;
-        if (size > int.MaxValue)
+        var numberOfEntries = searcher.NumberOfEntries;
+        
+        if (numberOfEntries == 0)
+            return 16;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:121]: why?

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 12]
```diff
@@ -113,16 +113,26 @@ protected static int LuceneGetPageSize(IndexSearcher searcher, long pageSize)
...
+
+        if (pageSize <= 0)
+            return (int)numberOfEntries;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:124]: What happens if we have > 2B entries, we'll get negative value here.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 13]
```diff
@@ -113,16 +113,26 @@ protected static int LuceneGetPageSize(IndexSearcher searcher, long pageSize)
...
-
-        if (size > bufferSize || query.Metadata.OrderBy is not null)
-            return (int)size;
-
-        return bufferSize;
+        
+        //If we have a binary operation, we need to pass a buffer large enough to hold all the individual results.
+        //We need to do this to get correct results and since we don't know how much results subqueries will have  we must create a buffer big enough to get all items from index
+        if (query.Metadata.OrderBy is not null || query.Metadata.IsDistinct || isBoolean)
+            return (int)numberOfEntries;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:132]: What happens if we have > 2B entries, we'll get negative value here.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs [hunk 14]
```diff
@@ -113,16 +113,26 @@ protected static int LuceneGetPageSize(IndexSearcher searcher, long pageSize)
...
+
+        var result = Math.Min(numberOfEntries, pageSize);
+        return (int)(result <= 0 ? 2 << 4 : result);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:135]: What happens if we have > 2B entries, we'll get negative value here.

File: src/Corax/IndexWriter.cs (hunk 7 above)
Comment (maciejaszyk) [reply src/Corax/IndexWriter.cs:311]: Deleted when implemented growable buffer.

File: src/Corax/Queries/MemoizationMatch.Provider.cs (hunk 8 above)
Comment (maciejaszyk) [reply src/Corax/Queries/MemoizationMatch.Provider.cs:9]: Probably IDE changed it and I've not noticed it. I will revert that

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs (hunk 12 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:124]: Fixed.

File: src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs (hunk 13 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/IndexOperationBase.cs:132]: Then it will be handled by condition before.

**Pull request merged**